File: scheduler/finish.py
"""
scheduler/finish.py — bitirme geçişi: son kartları derinlemesine ara.

Ana arama (tabu) çizelgenin tamamına yayılmış bir optimizasyon yapar ve
neredeyse tam bir sonuca çok hızlı varır. Ama son bir iki kartta tıkanır ve
orada kalır: v188'de 60 saniyeyle de 180 saniyeyle de aynı 283/285 çıkıyor —
yani sorun süre değil, aramanın o noktada yapabildiği hamlenin sınırlı olması.

Tabu araması her adımda TEK bir kartı oynatır. Son kartın yerleşebilmesi için
çoğu zaman üç dört kartın birlikte kayması gerekir ve böyle bir hamle, tek
kartlık adımların hiçbir dizilişinde ara durumu kötüleştirmeden görünmez.

Bu geçiş tam olarak o hamleyi arar ve yalnızca onu: açıkta kalan kart için
bütün yerleri dener, her yerde engelleyenleri söker, onları da özyinelemeli
olarak yeniden yerleştirir. Arama alanı küçüktür (bir iki kart), bu yüzden
tüketici arama burada hesaplıdır — bütün çizelgede olmazdı.

Sonuç, ana aramanın kurduğu çizelgeyi BOZMAZ: dal başarısızsa her şey birebir
eski hâline döner, başarılıysa yalnızca gereken kartlar yer değiştirmiş olur.
"""

from collections import defaultdict
from . import rules as R
from .model import subject_family


class Finisher:
    def __init__(self, world, rule_list, positions):
        self.w = world
        self.P = world.P
        self.D = world.D
        self.pos = list(positions)

        self.subject_once = any(r.kind == R.X_SUBJECT_ONCE_DAY and r.is_hard()
                                for r in rule_list)
        self.teacher_once = any(r.kind == R.X_TEACHER_ONCE_DAY and r.is_hard()
                                for r in rule_list)
        self.hard_adj = set()
        for r in rule_list:
            if r.kind == R.X_HARD_NOT_ADJACENT and r.is_hard():
                self.hard_adj |= set(r.subjects)

        n = len(world.cards)
        self.cls_cell = {}      # (sınıf, hücre) -> kart
        self.tch_cell = {}      # (öğretmen, hücre) -> kart
        self.fam_day = defaultdict(set)   # (sınıf, gün) -> {aile}
        self.tch_day = defaultdict(set)   # (sınıf, gün) -> {öğretmen}
        for i, idx in enumerate(self.pos):
            if idx >= 0:
                self._occupy(i, idx, True)

    def _occupy(self, i, idx, on):
        c = self.w.cards[i]
        d = idx // self.P
        for off in range(c.duration):
            cell = idx + off
            for ci in c.classes:
                if on: self.cls_cell[(ci, cell)] = i
                else: self.cls_cell.pop((ci, cell), None)
            if c.teacher >= 0:
                if on: self.tch_cell[(c.teacher, cell)] = i
                else: self.tch_cell.pop((c.teacher, cell), None)
        for ci in c.classes:
            key = (ci, d)
            if on:
                self.fam_day[key].add(c.family)
                if c.teacher >= 0: self.tch_day[key].add(c.teacher)
            else:
                # sayım gerektiği için yeniden kur
                self.fam_day[key] = {self.w.cards[j].family
                                     for (cj, cell), j in self.cls_cell.items()
                                     if cj == ci and cell // self.P == d}
                self.tch_day[key] = {self.w.cards[j].teacher
                                     for (cj, cell), j in self.cls_cell.items()
                                     if cj == ci and cell // self.P == d
                                     and self.w.cards[j].teacher >= 0}

    def _rules_ok(self, i, idx):
        c = self.w.cards[i]
        d, p = divmod(idx, self.P)
        for ci in c.classes:
            if self.subject_once and c.family in self.fam_day[(ci, d)]:
                return False
            if self.teacher_once and c.teacher >= 0 and c.teacher in self.tch_day[(ci, d)]:
                return False
            if c.subject in self.hard_adj:
                for nb in (p - 1, p + c.duration):
                    if 0 <= nb < self.P:
                        j = self.cls_cell.get((ci, d * self.P + nb))
                        if j is not None and self.w.cards[j].subject in self.hard_adj:
                            return False
        return True

    def _blockers(self, i, idx):
        c = self.w.cards[i]
        out = set()
        for off in range(c.duration):
            cell = idx + off
            for ci in c.classes:
                j = self.cls_cell.get((ci, cell))
                if j is not None: out.add(j)
            if c.teacher >= 0:
                j = self.tch_cell.get((c.teacher, cell))
                if j is not None: out.add(j)
        return out

    def _free(self, i, idx):
        return not self._blockers(i, idx)

    def place(self, i, depth, max_depth, banned):
        """Kartı yerleştirmeyi dener; gerekirse engelleyenleri taşır."""
        c = self.w.cards[i]
        # Önce boş bir yer
        for idx, _ in c.slots:
            if self._free(i, idx) and self._rules_ok(i, idx):
                self.pos[i] = idx; self._occupy(i, idx, True)
                return True
        if depth >= max_depth:
            return False
        # Sonra tahliyeyle
        for idx, _ in c.slots:
            if not self._rules_ok(i, idx):
                continue
            blk = self._blockers(i, idx)
            if not blk or len(blk) > 2 or (blk & banned):
                continue
            snap = list(self.pos)
            moved = []
            for j in blk:
                old = self.pos[j]
                self._occupy(j, old, False); self.pos[j] = -1
                moved.append((j, old))
            if self._free(i, idx) and self._rules_ok(i, idx):
                self.pos[i] = idx; self._occupy(i, idx, True)
                ok = True
                nb = banned | blk | {i}
                for j, _ in moved:
                    if not self.place(j, depth + 1, max_depth, nb):
                        ok = False; break
                if ok:
                    return True
            self._restore(snap)
        return False

    def _restore(self, snap):
        """Çizelgeyi anlık görüntüye birebir döndürür.

        Önce farklı olan her kart sökülür, sonra hepsi eski yerine konur; tek
        geçişte yapmak henüz sökülmemiş bir kartın üstüne yazma riski taşır.
        """
        for i, idx in enumerate(self.pos):
            if idx >= 0 and snap[i] != idx:
                self._occupy(i, idx, False); self.pos[i] = -1
        for i, idx in enumerate(snap):
            if idx >= 0 and self.pos[i] < 0 and self._free(i, idx):
                self.pos[i] = idx; self._occupy(i, idx, True)

    def run(self, max_depth=4):
        eksik = [i for i, idx in enumerate(self.pos)
                 if idx < 0 and self.w.cards[i].slots]
        kazanc = 0
        for i in sorted(eksik, key=lambda k: len(self.w.cards[k].slots)):
            if self.pos[i] >= 0:
                continue
            if self.place(i, 0, max_depth, frozenset()):
                kazanc += self.w.cards[i].duration * len(self.w.cards[i].classes)
        return self.pos, kazanc

File: web/src/engine/finish.ts
// scheduler/finish.py — bitirme geçişi: açıkta kalan son kartlar için
// tahliye zinciriyle derin arama. Başarısız dal birebir geri alınır.
import type { World } from './model.ts';
import { key } from './py.ts';
import * as R from './rules.ts';
import type { Rule } from './rules.ts';
import { validate } from './verify.ts';
import type { Pieces } from './verify.ts';

export class Finisher {
  w: World;
  P: number;
  pos: number[];
  rules: Rule[];
  forced: Set<string> | null;
  pieces: Pieces;
  subjectOnce: boolean;
  teacherOnce: boolean;
  subjectAdj: boolean;
  hardAdj = new Set<number>();
  clsCell = new Map<string, number>();
  tchCell = new Map<string, number>();
  famDay = new Map<string, Set<number>>();
  tchDay = new Map<string, Set<number>>();

  constructor(world: World, rules: Rule[], positions: number[], forced: Set<string> | null = null, pieces: Pieces | null = null) {
    this.w = world;
    this.P = world.P;
    this.pos = [...positions];
    this.rules = rules;
    this.forced = forced;
    this.pieces = new Map(pieces ?? []);
    this.subjectOnce = rules.some((r) => r.kind === R.X_SUBJECT_ONCE_DAY && r.isHard());
    this.teacherOnce = rules.some((r) => r.kind === R.X_TEACHER_ONCE_DAY && r.isHard());
    this.subjectAdj = rules.some((r) => r.kind === R.X_SUBJECT_NOT_ADJACENT && r.isHard());
    for (const r of rules) {
      if (r.kind === R.X_HARD_NOT_ADJACENT && r.isHard()) for (const s of r.subjects) this.hardAdj.add(world.subjectFamily[s]);
    }
    this.pos.forEach((idx, i) => { if (idx >= 0) this.occupy(i, idx, true); });
    for (const [i, cells] of this.pieces) {
      const c = world.cards[i];
      for (const cell of cells) {
        for (const ci of c.classes) {
          this.clsCell.set(key(ci, cell), -1 - i);
          this.day(this.famDay, ci, Math.floor(cell / this.P)).add(c.family);
          if (c.teacher >= 0) this.day(this.tchDay, ci, Math.floor(cell / this.P)).add(c.teacher);
        }
        if (c.teacher >= 0) this.tchCell.set(key(c.teacher, cell), -1 - i);
      }
    }
  }

  private day(m: Map<string, Set<number>>, ci: number, d: number): Set<number> {
    const k = key(ci, d);
    let s = m.get(k);
    if (!s) { s = new Set(); m.set(k, s); }
    return s;
  }

  private occupy(i: number, idx: number, on: boolean) {
    const c = this.w.cards[i];
    const d = Math.floor(idx / this.P);
    for (let off = 0; off < c.duration; off++) {
      const cell = idx + off;
      for (const ci of c.classes) {
        if (on) this.clsCell.set(key(ci, cell), i); else this.clsCell.delete(key(ci, cell));
      }
      if (c.teacher >= 0) {
        if (on) this.tchCell.set(key(c.teacher, cell), i); else this.tchCell.delete(key(c.teacher, cell));
      }
    }
    for (const ci of c.classes) {
      if (on) {
        this.day(this.famDay, ci, d).add(c.family);
        if (c.teacher >= 0) this.day(this.tchDay, ci, d).add(c.teacher);
      } else {
        const owners: number[] = [];
        for (const [k, j] of this.clsCell) {
          const [cj, cell] = k.split('\u0001').map(Number);
          if (cj === ci && Math.floor(cell / this.P) === d) owners.push(j < 0 ? -1 - j : j);
        }
        this.famDay.set(key(ci, d), new Set(owners.map((j) => this.w.cards[j].family)));
        this.tchDay.set(key(ci, d), new Set(owners.map((j) => this.w.cards[j].teacher).filter((t) => t >= 0)));
      }
    }
  }

  private rulesOk(i: number, idx: number): boolean {
    const c = this.w.cards[i];
    const d = Math.floor(idx / this.P), p = idx % this.P;
    for (const ci of c.classes) {
      if (this.subjectOnce && this.day(this.famDay, ci, d).has(c.family)) return false;
      if (this.teacherOnce && c.teacher >= 0 && this.day(this.tchDay, ci, d).has(c.teacher)) return false;
      if (this.hardAdj.has(c.family) || this.subjectAdj) {
        for (const nb of [p - 1, p + c.duration]) {
          if (nb < 0 || nb >= this.P) continue;
          const j = this.clsCell.get(key(ci, d * this.P + nb));
          if (j === undefined) continue;
          const o = this.w.cards[j < 0 ? -1 - j : j];
          if (this.hardAdj.has(c.family) && this.hardAdj.has(o.family) && o.family !== c.family) return false;
          if (this.subjectAdj && o.family === c.family) return false;
        }
      }
    }
    return true;
  }

  private blockers(i: number, idx: number): Set<number> {
    const c = this.w.cards[i];
    const out = new Set<number>();
    for (let off = 0; off < c.duration; off++) {
      const cell = idx + off;
      for (const ci of c.classes) {
        const j = this.clsCell.get(key(ci, cell));
        if (j !== undefined) out.add(j);
      }
      if (c.teacher >= 0) {
        const j = this.tchCell.get(key(c.teacher, cell));
        if (j !== undefined) out.add(j);
      }
    }
    return out;
  }

  private free(i: number, idx: number): boolean {
    return this.blockers(i, idx).size === 0;
  }

  private place(i: number, depth: number, maxDepth: number, banned: Set<number>): boolean {
    const c = this.w.cards[i];
    for (const idx of c.slots) {
      if (this.free(i, idx) && this.rulesOk(i, idx)) {
        this.pos[i] = idx;
        this.occupy(i, idx, true);
        return true;
      }
    }
    if (depth >= maxDepth) return false;
    for (const idx of c.slots) {
      if (!this.rulesOk(i, idx)) continue;
      const blk = [...this.blockers(i, idx)].sort((a, b) => a - b);
      if (!blk.length || blk.length > 2 || blk.some((j) => banned.has(j)) || blk.some((j) => j < 0)
        || blk.some((j) => this.w.cards[j].lockedAt !== null)) continue;   // kilitli ders asla tahliye edilmez
      const snap = [...this.pos];
      const moved: number[] = [];
      for (const j of blk) {
        this.occupy(j, this.pos[j], false);
        this.pos[j] = -1;
        moved.push(j);
      }
      if (this.free(i, idx) && this.rulesOk(i, idx)) {
        this.pos[i] = idx;
        this.occupy(i, idx, true);
        let ok = true;
        const nb = new Set([...banned, ...blk, i]);
        for (const j of moved) {
          if (!this.place(j, depth + 1, maxDepth, nb)) { ok = false; break; }
        }
        if (ok) return true;
      }
      this.restore(snap);
    }
    return false;
  }

  private restore(snap: number[]) {
    this.pos.forEach((idx, i) => {
      if (idx >= 0 && snap[i] !== idx) { this.occupy(i, idx, false); this.pos[i] = -1; }
    });
    snap.forEach((idx, i) => {
      if (idx >= 0 && this.pos[i] < 0 && this.free(i, idx)) { this.pos[i] = idx; this.occupy(i, idx, true); }
    });
  }

  run(maxDepth = 4): [number[], number] {
    const eksik = this.pos.map((idx, i) => [idx, i] as const)
      .filter(([idx, i]) => idx < 0 && this.w.cards[i].slots.length && !this.pieces.has(i)).map(([, i]) => i);
    let kazanc = 0;
    for (const i of [...eksik].sort((a, b) => this.w.cards[a].slots.length - this.w.cards[b].slots.length)) {
      if (this.pos[i] >= 0) continue;
      const snap = [...this.pos];
      if (this.place(i, 0, maxDepth, new Set())) {
        const [errs] = validate(this.w, this.rules, this.pos, true, this.forced, this.pieces);
        if (errs.length) { this.restore(snap); continue; }
        kazanc += this.w.cards[i].duration * this.w.cards[i].classes.length;
      }
    }
    return [this.pos, kazanc];
  }
}
